### ISSUE TITLE
load: allow loading relative paths of arbitrary extension

### ISSUE THREAD

**Issue by eedwards-sk** [bats-core/bats-core#79]
I'm not sure what the goal of expecting a helper script to be an extension of `.bash` is, but it makes absolutely zero sense.

Let me specify a path and load that file. Don't force the extension on me.

**Comment by tterranigma** [comment]
Why don't you just use `source`? `load` is a wrapper around `source` to resolve the path to the script automatically.

**Comment by eedwards-sk** [comment]
@tterranigma

I'm already using source now, it has problems too, mostly #72

Generally, if a product has a feature for a purpose, I try to use it. Thus, the request to allow specific files :)

I want to support relative pathways, as well. The path I would pass into `load` would be absolute sometimes and relative others, depending on the `make` lifecycle step I'm currently running it from.

**Comment by tterranigma** [comment]
I also do not like having to name my files with `.bash` at the end. But, then, this issue should better be named 'load: allow loading relative paths of arbitrary extension` which also covers 'no extension'.

**Comment by eedwards-sk** [comment]
Good point. I'll update the title.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by jasonkarns** [comment]
I don't know sam stephenson's rationale for originally creating `load` to use `.bash`, but I'm personally pretty strongly in favor of it. It reinforces the convention that executable shell scripts are extension-less, and sourced files are extension-ed. (totally a word) And this is _ Bash_ Automated Testing System after all, so tests helpers would be written in bash, not any other shell variant.

Right now, there are alternatives (`source`) for those who don't want the forced extension. But if we removed it, there would be no alternative for people to opt into the convention. (I don't think I'd consider any scheme to "add `.bash` if missing `.`" an acceptable hack.)

But, that's just, like, my opinion.

![](https://media.giphy.com/media/F3G8ymQkOkbII/giphy.gif)

**Comment by eedwards-sk** [comment]
@jasonkarns 

>but I'm personally pretty strongly in favor of it

Forcing a convention for religious regions is silly. You're basically making a tabs vs spaces argument. While I love arguing over things that don't matter, it would be a lot easier to just fix the problem.

Since you can actually source whatever file you want, extension or not, forcing an extension makes no sense.

Also, there's no advantage to forcing the extension, only disadvantages, and the fix is to remove the line that adds `.bash` to the name of the file you specify to `load`. Extremely simple fix.

If we're going to address #72 with #80 then this forced extension becomes a bigger problem.

**Comment by mbland** [comment]
@eedwards-sk You're raising a valid point, but tone down the rhetoric. We're here to have a respectful conversation, and @jasonkarns merely explained his preference and clearly stated that it's merely his opinion.

**Comment by eedwards-sk** [comment]
@mbland Understood, sorry for that!

**Comment by mbland** [comment]
@eedwards-sk No worries. I also tend to agree with your view. While I see @jasonkarns point about wanting to promote a convention to distinguish executables from modules (if you will), I tend to think the `load` command isn't the right vehicle to promote such a convention.

**Comment by sublimino** [comment]
FWIW I'd prefer the uniformity of `.sh` on my scripts, but I can't really justify it - again just personal preference. 

And I also understand @jasonkarns' position, although I'd never used `source` instead and just renamed the scripts to `.bash` to comply.

**Comment by jasonkarns** [comment]
Any chance we could get @sstephenson's original thinking on this matter?

**Comment by cpmills1975** [comment]
I'd prefer not to have .bash enforced for the following reason: I'm working on a project that has years of bash scripting behind it. I'm trying to encourage TDD using bats, rather than any of the other myriad of test engines, because of the project background in bash. Trying to get bash developers to start writing tests in rspec or ruby or even python simply isn't going to encourage TDD.

I'm sourcing a couple of files that do not currently end in .bash and changing the extension on these files would cause a whole bag of hurt elsewhere in the project. The files define a whole raft of constants and functions used throughout our scripts.

If it were possible to include the content of these files in-line, that would be truly awesome - the constants file particularly contains a whole bunch of bash associative arrays and I understand that due to a bash limitation, an associative array defined in a function cannot be accessed outside of that function. That isn't currently a problem in our existing scripts as we invariably source the constants file at the top of the scripts we write.

**Comment by jasonkarns** [comment]
@eedwards-sk do you have a known example or use case for loading something that isn't a test helper? I'm struggling to imagine a scenario where it loads something other than a test helper, that doesn't also completely invalidate the value of the test.

The only thing I can think of is to load scripts that are also typically sourced in prod. Though for the test to have value at that point, one would just be invoking individual functions. Which, IMO, is such a low-value exercise that it makes the test almost meaningless. (Unless one is loading the scripts as part of the test, and checking for side effects. But woo boy, that's a rat's nest.)

**Comment by eedwards-sk** [comment]
@jasonkarns I'm writing a bash library and I load the library as part of setup

Also, even if it's only used for test helpers, it doesn't necessitate the extension

**Comment by bvberkum** [comment]
The current solution is okay, but the implementation is fairly rigid. A bit too much.

I am now moving my ``*bats`` files into subdirectories, and I want more control over de PWD for special CI or other script setups. Right now, scripts break simply because I need to rewrite:
```
load assert
load ../assert
```

I don't know wether these files will stay there, or wether some other project will have other requirements. The altered statements add nothing to the test.
The hardcoded '.bash' extension is equally unwanted.

TLDR;

I'm going to make a PR for a proposal with this heuristic:
go over BATS_TEST_LIB in sequence and return on first existing name.
```
load() # ( PATH | NAME )
{
  test -e "$1" && {
    . "$1"
    return $?
  }
  for i in ${BATS_TEST_LIB//:/ }
  do
      test -e "$i/$1" && {
        . "$i/$1"
        return $?
      }
      test -e "$i/$1.bash" && {
        . "$i/$1.bash"
        return $?
      }
      test -e "$i/$1.sh" && {
        . "$i/$1.sh"
        return $?
      }
  done
}
```
Should probably add ``BATS_LIB_EXTS=bash\ sh`` as well

**Comment by jasonkarns** [comment]
@mbland It was a bit of a stretch on my part to say that testing functions directly is lower value. (If the thing under test is literally a library of functions, then that's precisely the desired goal.)

However, if the original bats readme has any possible elucidation of Stephenson's thinking on the matter, I think it pretty clearly focuses on `load` being used to load test helpers, fixtures, test setup, etc; as opposed to loading functions under test. In that vein, I don't see any problem with the added friction when `load` is used to source other files.

On the other, other hand, I'm okay pushing `load` to be more useful to a broader group. As long as it still provides value above and beyond `source`.

@bvberkum I'm mostly okay with the spirit of your suggestion above. But I'll add two additional patterns that could use some love with that implementation:

1. loading bats helper libs distributed as npm modules (`load "../node_modules/bats-assert/load.bash"`)
2. the pattern introduced by @ztombol wherein the main entrypoint for a "library" is `load.bash` https://github.com/ztombol/bats-docs#loading

Perhaps beyond searching the test directory for `foo{,.sh,.bash}` files, we could leverage `load` to provide a common entrypoint type of functionality. For instance: if the path provided is a directory, source the `load{,.sh,.bash}` file (or `main{,.sh,.bash}` or `index{,.sh,.bash}`, whatever)

I'd also like to make `load` work well with [basher](https://github.com/basherpm/basher) packages. Basher has its own `include` helper, but it doesn't seem to define a common entrypoint filename. (`include` just sources files relative to the package's directory).

**Comment by bvberkum** [comment]
@jasonkarns 
Ty for the input. I think there should be simple conventions underlaying a routine. So I assume BATS_LIB_PATH is prepared and static at the point of Bats execution, ie. add node_modules etc. first.

On the other hand possibly we can combine two or more variables, a lookup path and a local path set. I have something in a PoC with two for loops but that seemed to convoluted. Like this:

```
BATS_LIB_PATH=/project1:/project:/lib1
# Test each lib path for existing basedir:
BATS_LIB_BASE=helper:test/helper:vendor/xyz/lib:node_modules/bats-abcd

BATS_LIB_EXSTS=sh\ bash
```

Add another for the ztombol module load scenario:
```
BATS_LIB_DEFAULT=load
```

**Comment by bvberkum** [comment]
ToDo's
- more extensive testing and specs for behaviour
- review for more optimum way to determine default BATS_LIB_PATH; with BATS_TEST_DIR
- Rewrite to bash (ie def. var values)

**Comment by Yunir** [comment]
Why this is still open - for about 1 year. Is this project still alive?

**Comment by sublimino** [comment]
@Yunir Hi! Yes, we're still alive, but PRs with ToDO's (like this one) won't get merged as they're incomplete. 

Perhaps @bvberkum has an update -- although I'd guess with the time passed this PR may require another contributor to get it over the line.

**Comment by ntnn** [comment]
@sublimino Well, tbh - even PRs without TODOs aren't getting merged. Mine has been open since 2017 with one comment from @mbland and no reaction from assigned reviewers.
https://github.com/bats-core/bats-core/pull/27

**Comment by sublimino** [comment]
Thanks for the note @ntnn -- the merges in the last 6 months or so have been some major refactors, so most old PRs don't merge without work. I haven't been picking up PRs that I haven't been involved with, and the other maintainers haven't been around for a while. 

I have committed in https://github.com/bats-core/bats-core/pull/153 to await a refactoring of that PR into the current codebase from @martin-schulze-vireso. When that is done we will need to take stock and review everything for 1.2.

That said I'm happy to help get these over the line as I agree they're both useful. I've created a [1.2 project board](https://github.com/bats-core/bats-core/projects/4?add_cards_query=is%3Aopen) to track these. There will be a rebase required once https://github.com/bats-core/bats-core/pull/153 is merged.
